econsa/correlation.py

```diff
@@ -0,0 +1,175 @@
+"""Map arbitrary correlation matrix to Gaussian.
+
+This module implements methods from two papers to map arbitrary correlation
+matrix into correlation matrix for Gaussian copulas.
+"""
+import chaospy as cp
+import numpy as np
+from scipy import optimize
+from statsmodels.stats.correlation_tools import corr_nearest
+
+
+def gc_correlation(marginals, corr, check_corr=True, force_calc=False):
+    """Correlation for Gaussian copula.
+
+    This function implements the algorithm outlined in Section 4.2 of [K2012]_
+    to map arbitrary correlation matrix to an correlation matrix for
+    Gaussian copula.
+    For special combination of distributions, use the values from Table 4. of
+    [L1986]_.
+
+    Since chaospy's copula functions only accept positive definite correlation matrix,
+    this function also checks the output,
+    and transforms to nearest positive definite matrix if it is not already.
+
+    Parameters
+    ----------
+    marginals : chaospy.distributions
+        Marginal distributions of the correlated variables. All marginals must be chaospy
+        distributions, otherwise returns error.
+
+    corr : array_like
+        The correlation matrix to be transformed.
+
+    check_corr : bool, optional
+        Check that `corr` is symmetric, all elements are between 0 and 1,
+        all diagonal elements are 1,
+        and all eigenvalue is positive (default value is `True`).
+
+    force_calc : bool, optional
+        When `True`, calculate the covariances ignoring all special preset combinations of marginals
+        (default value is `False`).
+
+    Returns
+    -------
+    gc_corr : numpy.ndarray
+        The transformed correlation matrix that is ready to be fed into a Gaussian copula.
+
+    References
+    ----------
+    .. [K2012] Kucherenko, S., Tarantola, S., & Annoni, P. (2012).
+        Estimation of global sensitivity indices for models with
+        dependent variables. Computer Physics Communications, 183(4), 937–946.
+
+    .. [L1986] Liu, P.-L., & Der Kiureghian, A. (1986).
+        Multivariate distribution models with prescribed marginals
+        and covariances. Probabilistic Engineering Mechanics, 1(2), 105–112.
+
+    Examples
+    --------
+    >>> corr = [[1.0, 0.6, 0.3], [0.6, 1.0, 0.0], [0.3, 0.0, 1.0]]
+    >>> marginals = [cp.Normal(1.00), cp.Uniform(lower=-4.00), cp.Normal(4.20)]
+    >>> corr_transformed = gc_correlation(marginals, corr)
+    >>> copula = cp.Nataf(cp.J(*marginals), corr_transformed)
+    >>> corr_copula = np.corrcoef(copula.sample(1000000))
+    >>> np.testing.assert_almost_equal(corr, corr_copula, decimal=6)
+    """
+    corr = np.atleast_2d(corr)
+
+    # Test that marginals are all cp.distributions.
+    for marginal in marginals:
+        if issubclass(type(marginal), cp.distributions.baseclass.Dist):
+            continue
+        else:
+            raise NotImplementedError("marginals must be chaospy distributions")
+
+    if check_corr:
```

Comment:
Let's just always run those checks and get rid of the optional argument `check_corr`.